libraries/Microsoft.Bot.Builder/ActivityHandler.cs

```diff
@@ -706,7 +729,69 @@ protected virtual Task OnUnrecognizedActivityTypeAsync(ITurnContext turnContext,
             return Task.CompletedTask;
         }
 
-        private AdaptiveCardInvokeValue GetAdaptiveCardInvokeValue(IInvokeActivity activity)
+        private static SearchInvokeValue GetSearchInvokeValue(IInvokeActivity activity)
+        {
+            if (activity.Value == null)
+            {
+                var response = CreateAdaptiveCardInvokeErrorResponse(HttpStatusCode.BadRequest, "BadRequest", "Missing value property for search");
+                throw new InvokeResponseException(HttpStatusCode.BadRequest, response);
+            }
+
+            var obj = activity.Value as JObject;
+            if (obj == null)
+            {
+                var response = CreateAdaptiveCardInvokeErrorResponse(HttpStatusCode.BadRequest, "BadRequest", "Value property is not properly formed for search");
+                throw new InvokeResponseException(HttpStatusCode.BadRequest, response);
+            }
+
+            SearchInvokeValue invokeValue = null;
+            
+            try
+            {
+                invokeValue = obj.ToObject<SearchInvokeValue>();
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                var errorResponse = CreateAdaptiveCardInvokeErrorResponse(HttpStatusCode.BadRequest, "BadRequest", "Value property is not valid for search");
+                throw new InvokeResponseException(HttpStatusCode.BadRequest, errorResponse);
+            }
+
+            ValidateSearchInvokeValue(invokeValue, activity.ChannelId);
+            return invokeValue;
+        }
+
+        private static void ValidateSearchInvokeValue(SearchInvokeValue searchInvokeValue, string channelId)
+        {
+            string missingField = null;
+
+            if (string.IsNullOrEmpty(searchInvokeValue.Kind))
+            {
+                // Teams does not always send the 'kind' field. Default to 'search'.
+                if (Connector.Channels.Msteams.Equals(channelId, StringComparison.OrdinalIgnoreCase))
+                {
+                    searchInvokeValue.Kind = SearchInvokeTypes.Search;
+                }
+                else
+                {
+                    missingField = nameof(searchInvokeValue.Kind);
+                }
+            }
+
+            if (string.IsNullOrEmpty(searchInvokeValue.QueryText))
+            {
+                missingField = nameof(searchInvokeValue.QueryText);
```

Comment:
nit: isn't nameof going to give you the C# identifier whereas what you really want in the error message is the JSON property name. OK this is perhaps a nit because the only difference in this case at the current time is the casing.